add garden size estimate evaluation notebook

crispy-wonton opened the pull request:
Fixes #114 

# Description

add garden size estimate evaluation notebook

# Instructions for Reviewer

to generate the notebook, ensure `jupytext` is installed and then run the following line in terminal:
`jupytext --to notebook asf_heat_pump_suitability/analysis/garden_size_estimates/20250106_evaluate_garden_size_estimates.py`

Please pay special attention to ...
Please could you provide some feedback on this analysis. Namely:
- Do you think the selection criteria are appropriate and applied in the correct order?
- Have I accounted for the appropriate corrections to make this a fair analysis (e.g. applying weights, checking weights add to at least 0.9 etc.)
- Are there any suggestions you have to make the analysis more robust or accurate?
- Have the weights been applied correctly?
- Is there anything we should add to the analysis?

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
add garden size estimate evaluation notebook

crispy-wonton committed:
update evaluation data file name in `20250106_evaluate_garden_size_estimates.py`

lizgzil commented on asf_heat_pump_suitability/analysis/garden_size_estimates/20250106_evaluate_garden_size_estimates.py:
```diff
@@ -0,0 +1,385 @@
+
+# %%
+# Filter to overall average garden size. We need to compare all property types to the total average to be able to use the weights
+total_avg_gardens_df = avg_gardens_df.filter(
+    pl.col("msoa_avg_outdoor_space_property_type") == "unknown"
+)
```
How come you didn't compare all the property types?

crispy-wonton replied:
Here I didn't compare for property types because the dataset is weighted on property type, so I didn't think I can use the weights whilst also filtering for property type because the weights for each property type will add to less than 1. I think this would mean the resulting weighted means would be weighted to the proportion of property type in the LSOA.

However, now I think about it again, I think I could multiply the weights by the number of households per LSOA and calculate the average that way. I will test it out.

lizgzil commented on asf_heat_pump_suitability/analysis/garden_size_estimates/20250106_evaluate_garden_size_estimates.py:
```diff
@@ -0,0 +1,385 @@
+)
+
+# %%
+# Get counts of UPRN per garden
+uprn_count = gardens_df.group_by("NATIONALCADASTRALREFERENCE").agg(
+    pl.col("UPRN").count().alias("UPRN_count")
+)
+
+# Assign 1 garden size per cadastral
+cadastral_garden_size = gardens_df.group_by("NATIONALCADASTRALREFERENCE").agg(
+    pl.col("garden_area_m2").first().alias("cadastral_garden_size_m2")
+)
+
+# Join to garden size df
+gardens_df = gardens_df.join(
+    uprn_count, how="left", on="NATIONALCADASTRALREFERENCE"
+).join(cadastral_garden_size, how="left", on="NATIONALCADASTRALREFERENCE")
+
+# Divide shared gardens equally among UPRNs sharing gardens
+gardens_df = gardens_df.with_columns(
+    (pl.col("cadastral_garden_size_m2") / pl.col("UPRN_count")).alias(
+        "divided_garden_area_m2"
+    )
+)
```
I wasn't expecting these steps. I thought "garden_area_m2" was the estimation for each UPRN, so surprised "divided_garden_area_m2" needed to be calculated. Where have I misunderstood?

crispy-wonton replied:
This is a good question. The reason is:

For properties that get joined to the same garden (i.e. they share a garden, or at least outdoor space) we decided in a previous sprint review that we would assign all properties which share a garden the overall size of the garden. E.g. if 10 houses share a 100m2 garden, they will all be assigned 100m2 garden size. The logic was that if a building has the appropriate outdoor space, we consider that an apartment within the building has appropriate space in theory.

I had to conduct this extra division step which partitions the size equally among households. E.g. the 10 houses sharing a 100m2 garden will now each get assigned 10m2. I had to do this to avoid double counting gardens when calculating the average. This is flawed, because if there are apartments from the same building missing from the EPC dataset, we will overestimate the divided garden size. E.g. if we only have 5 of the 10 houses sharing the 100m2 garden in our dataset, they would get assigned 20m2 each instead of 10m2.

lizgzil commented on asf_heat_pump_suitability/analysis/garden_size_estimates/20250106_evaluate_garden_size_estimates.py:
```diff
@@ -0,0 +1,385 @@
+    pl.col("divided_garden_area_m2").is_not_null(),
+    pl.col("divided_garden_area_m2") > 0,
+    pl.col("divided_garden_area_m2") < pl.col("divided_garden_area_m2").quantile(0.99),
+    pl.col("weight").is_not_null(),
```
I can't remember why a weight would be null? Was it when the value was from a dummy property?

crispy-wonton replied:
Weight can be null if:
- The UPRN did not get matched to an LSOA/DZ code
- There was no census data found for the LSOA/DZ code
- The EPC record did not have the required datapoints for reweighting (property type, tenure, & build year [England and Wales only])

There should not be any dummy rows in the dataset we are using here.

lizgzil commented on asf_heat_pump_suitability/analysis/garden_size_estimates/20250106_evaluate_garden_size_estimates.py:
```diff
@@ -0,0 +1,385 @@
+)
+
+# Filter results to MSOAs with averages calculated from 15 or more properties
+results = results.filter(pl.col("n_properties") >= 15)
```
I think here and in the weighted results you could increase this threshold - perhaps set to the 0.1 quantile? (which I think is 51). or was there a reason to use 15?

crispy-wonton replied:
There wasn't a particular reason behind 15, I think, but we chose a deliberately low number to maximise the number of LSOAs that would be weighted.

For comparing with the average I think it makes sense to increase to 51. We can try it with the weighting as well but bear in mind we will end up with more unweighted LSOA scores.

lizgzil reviewed (commented):
Hey @crispy-wonton this looked good to me, and I thought the filtering steps made sense (and in the right order) - I commented on a few bits for clarification though.

crispy-wonton committed:
update evaluate_garden_size_estimates.py notebook analysis
- add alternative deduplication step into notebook for replicability
- add documentation to explain alternative deduplication methodology
- assess weighted results by property type

lizgzil commented on asf_heat_pump_suitability/analysis/garden_size_estimates/20250106_evaluate_garden_size_estimates.py:
```diff
@@ -17,11 +28,43 @@
 # %% [markdown]
-# ### 2. Load garden size estimates
+# ### 2. Load garden size estimates and deduplicate
+#
+# For properties that get joined to the same garden (i.e. they share a garden, or at least outdoor space) we decided in a sprint review that we would assign all properties which share a garden the overall size of the garden. E.g. if 10 houses share a 100m2 garden, they will all be assigned 100m2 garden size. The logic was that if a building has the appropriate outdoor space, we consider that an a ...
+#
+# For this analysis, we have to conduct this extra division step which partitions the garden size equally among households who share a garden. E.g. the 10 houses sharing a 100m2 garden will now each get assigned 10m2. This is required to avoid double counting gardens when calculating the average and to make the resulting garden size estimates comparable to the ONS MSOA-level averages which are ca ...
+#
+# A caveat of our deduplication here is that if there are apartments from the same building missing from the EPC dataset or missing from our garden size estimates (e.g. because they are missing valid UPRN or there are data gaps in the Microsoft buildings / INSPIRE land parcel datasets), we will overestimate the divided garden size. E.g. if we only have records for 5 of the 10 houses sharing the 1 ...
```
super useful to have this here!

lizgzil reviewed (approved):
Thanks @crispy-wonton - looks good!
